src/email-providers/class-sendgrid.php

```diff
@@ -0,0 +1,464 @@
+		}
+
+		$css_to_inline_styles = new CssToInlineStyles();
+		$html_content         = $css_to_inline_styles->convert(
+			$html_content,
+			''
+		);
+
+		$text_content = wp_strip_all_tags( $html_content );
+		$subject      = get_post_meta( $newsletter_id, 'nb_newsletter_subject', true );
+
+		$request_body               = (object) [];
+		$request_body->name         = get_the_title( $newsletter_id );
+		$request_body->categories   = [];
+		$request_body->email_config = (object) [];
+		$request_body->send_to      = (object) [];
+		// $request_body->email_config->custom_unsubscribe_url = '';
+		$request_body->email_config->html_content           = $html_content;
+		$request_body->email_config->ip_pool                = null;
+		$request_body->email_config->plain_content          = $text_content;
+		$request_body->email_config->generate_plain_content = true;
+		$request_body->email_config->sender_id              = $sender_id ?? 0;
+		$request_body->email_config->subject                = $subject;
+		$request_body->email_config->suppression_group_id   = get_post_meta( $newsletter_id, 'nb_newsletter_suppression_group', true );
+
+		$request_body->send_to->list_ids    = $list_ids;
+		$request_body->send_to->segment_ids = [];
+
+		$sg = $this->get_client();
+		if ( empty( $sg ) ) {
+			return false;
+		}
+		$response = $sg->client->marketing()->singlesends()->post( $request_body );
+
+		return [
+			'response'         => $response->body(),
+			'http_status_code' => $response->statusCode(),
+		];
+	}
+
+	/**
+	 * Sends a campaign.
+	 *
+	 * @param string $campaign_id The campaign id.
+	 * @return array{
+	 *   response: mixed,
+	 *   http_status_code: int,
+	 * }|false  The response from the API.
+	 */
+	public function send_campaign( string $campaign_id ): array|false {
+		$sg = $this->get_client();
+		if ( empty( $sg ) ) {
+			return false;
+		}
+
+		try {
+			// @phpstan-ignore-next-line
+			$response = $sg->client->marketing()->singlesends()->_( $campaign_id )->schedule()->put( [ 'send_at' => 'now' ] );
+			return [
+				'response'         => $response->body(),
+				'http_status_code' => $response->statusCode(),
+			];
+		} catch ( \Exception $ex ) {
+			return false;
+		}
+	}
+
+	/**
+	 * Gets campaign summary.
+	 *
+	 * @param string $campaign_id The campaign id.
+	 * @return array{
+	 *   response: mixed,
+	 *   http_status_code: int,
+	 * }|false  The response from the API.
+	 *
+	 * @todo: Get recipients, total opened, unique opened.
+	 */
+	public function get_campaign_summary( string $campaign_id ): array|false {
+		$sg = $this->get_client();
+		if ( empty( $sg ) ) {
+			return false;
+		}
+		$response = $sg->client->marketing()->singlesends()->_( $campaign_id )->get();
+		$body     = (object) json_decode( $response->body() );
+		return [
+			'response'         => [
+				'Status'       => $body->status,
+				'Name'         => $body->name,
+				'Recipients'   => 'N/A',
+				'TotalOpened'  => 'N/A',
+				'UniqueOpened' => 'N/A',
```

Comment:
Is N/A a value we ingest elsewhere? Should it be a const?

Reply:
ideally we'll get this data from somewhere, but I'm not sure where it lives yet.